lib/Dialect/SV/SVOps.cpp

```diff
@@ -1944,6 +1946,304 @@ void BindOp::build(OpBuilder &builder, OperationState &odsState, StringAttr mod,
   odsState.addAttribute("instance", ref);
 }
 
+//===----------------------------------------------------------------------===//
+// SVVerbatimSourceOp
+//===----------------------------------------------------------------------===//
+
+void SVVerbatimSourceOp::print(OpAsmPrinter &p) {
+  p << ' ';
+
+  StringRef visibilityAttrName = SymbolTable::getVisibilityAttrName();
+  if (auto visibility = (*this)->getAttrOfType<StringAttr>(visibilityAttrName))
+    p << visibility.getValue() << ' ';
+
+  p.printSymbolName(getSymName());
+
+  // Print parameters
+  circt::printOptionalParameterList(p, *this, getParameters());
+
+  // Print attributes using the helper function
+  SmallVector<StringRef> omittedAttrs = {SymbolTable::getSymbolAttrName(),
+                                         "parameters", visibilityAttrName};
+
+  p.printOptionalAttrDictWithKeyword((*this)->getAttrs(), omittedAttrs);
+}
+
+ParseResult SVVerbatimSourceOp::parse(OpAsmParser &parser,
+                                      OperationState &result) {
+
+  // parse optional visibility
+  StringRef visibilityAttrName = SymbolTable::getVisibilityAttrName();
+  StringRef visibility;
+  if (succeeded(parser.parseOptionalKeyword(&visibility,
+                                            {"public", "private", "nested"}))) {
+    result.addAttribute(visibilityAttrName,
+                        parser.getBuilder().getStringAttr(visibility));
+  }
+
+  // Parse the symbol name
+  StringAttr nameAttr;
+  if (parser.parseSymbolName(nameAttr, SymbolTable::getSymbolAttrName(),
+                             result.attributes))
+    return failure();
+
+  // Parse optional parameters
+  ArrayAttr parameters;
+  if (circt::parseOptionalParameterList(parser, parameters))
+    return failure();
+  result.addAttribute("parameters", parameters);
+
+  // Parse attributes using the helper function
+  if (parser.parseOptionalAttrDictWithKeyword(result.attributes))
+    return failure();
+
+  return success();
+}
+
+LogicalResult SVVerbatimSourceOp::verify() {
+  // must have verbatim content
+  if (getContent().empty())
+    return emitOpError("missing or empty content attribute");
+
+  return success();
+}
```

Comment:
This is fine. There's a slightly cleaner way to define a custom attribute constraint of "non empty string" and then using this in ODS. Follow-up work.